dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Voltage/TCL.mo

```diff
@@ -23,62 +23,74 @@ model TCL "Tap Changer Lock (TCL)"
 
     Connectors.ImPin UMonitored (value (unit = "kV")) "Monitored voltage";
-    Connectors.BPin lockOrder (value (start = locked0)) "TCL lock order";
+    Connectors.BPin lockOrder (value (start = locked0)) "TCL manual lock order";
+    Connectors.BPin unlockOrder (value (start = locked0)) "TCL manual lock order";
     Boolean lockedT (start = locked0) "High voltage transformers locked ?";
     Boolean lockedD (start = locked0) "Low voltage transformers locked ?";
 
   protected
 
     parameter Boolean locked0 = false "Is the TCL initially locked ?";
+    parameter State state0 = State.Standard "Initial state";
 
     Boolean UUnderMin (start = false) "U < Umin ?";
     Types.Time tUnderUmin (start = Constants.inf) "Time when U < Umin";
     Types.Time tLocked (start = Constants.inf) "Time when the TCL was locked";
-    Boolean locked (start = locked0) "TCL locked ?";
+    State state(start = state0) "State of the automata throughout the simulation";
 
   equation
 
     // Check when the monitored voltage goes below UMin
-    // If the TCL is manually locked we ignore the voltage related events. Might be reviewed later.
-    when UMonitored.value < UMin and UMonitored.value > 0  and not (lockOrder.value) then
+    when UMonitored.value < UMin then
       UUnderMin = true;
       tUnderUmin = time;
-      Timeline.logEvent1(TimelineKeys.TapChangersArming);
-    elsewhen UMonitored.value > UMin and pre(UUnderMin) and not pre(lockOrder.value) then
+    elsewhen UMonitored.value >= UMin and pre(UUnderMin) then
       UUnderMin = false;
       tUnderUmin = Constants.inf;
-      Timeline.logEvent1(TimelineKeys.TapChangersUnarming);
     end when;
 
-    // Lock order activation
-    when ( (UUnderMin and time - tUnderUmin >= tLagBeforeLocked) or lockOrder.value ) and (not pre(locked)) then
-      locked = true;
+    // Automata state transitions
+    // Note that the deactivation from the Locked state is only manual
+    when UUnderMin and pre(state) == State. Standard then
+      state = State.Armed;
+      tLocked = Constants.inf;
+      Timeline.logEvent1(TimelineKeys.TapChangersArming);
+    elsewhen not(UUnderMin) and pre(state) == State.Armed then
+      state = State.Standard;
+      tLocked = Constants.inf;
+      Timeline.logEvent1(TimelineKeys.TapChangersUnarming);
+    elsewhen (((time - tUnderUmin >= tLagBeforeLocked) or (lockOrder.value and not(pre(lockOrder.value)))) and pre(state) == State.Armed) or (lockOrder.value and not(pre(lockOrder.value)) and pre(state) == State.Standard) then
+      state = State.Locked;
       tLocked = time;
       Timeline.logEvent1(TimelineKeys.TapChangersLocked);
-    elsewhen not(UUnderMin) and not(lockOrder.value) and (pre(locked)) then
-      locked = false;
+    elsewhen unlockOrder.value and not(pre(unlockOrder.value)) and pre(state) == State.Locked then
+      state = State.Standard;
       tLocked = Constants.inf;
       Timeline.logEvent1(TimelineKeys.TapChangersUnlocked);
     end when;
```

Comment:
From my point of view in the lockOrder and unlockOrder based transitions, you don't really need the three conditions: current lock/unlockOrder.value, pre(lockOrder.value) and pre(state).

Indeed, for example, if pre(lockOrder.value) = true, then pre(state) is locked by construction.

It is also visible in the next when loops when you don't use the pre(state) to unlock.

It would lead to:
((time - tUnderUmin >= tLagBeforeLocked) or lockOrder.value) and pre(state) == State.Armed) or (lockOrder.value and pre(state) == State.Standard) then

and

elsewhen unlockOrder.value and pre(state) == State.Locked then

except if I have missed something which is possible ;)

Reply:
What you propose doesn't work unfortunately, I have to catch the change in lockOrder and unlockOrder only, otherwise the conditions can remain true and the automata cyclically change of state.

Reply:
I found in a documentation that the unlock condition must also verify that the arming condition is false so I updated a bit the model. The condition for unlock becomes: 

`elsewhen unlockOrder.value and not(pre(unlockOrder.value)) and not(UUnderMin) and pre(state) == State.Locked then`